File: src/DashSpec.Core/Parsing/DashboardShellParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal enum DashboardShellMode
{
    DashboardBody,
    TabModuleStandalone,
    TabModuleEmbedded,
}

internal sealed class DashboardShellContext
{
    public required DashboardShellMode Mode { get; init; }

    public string? SpecDirectory { get; init; }

    public string? TabModuleId { get; init; }

    public IReadOnlyList<FilterDefinition>? ParentFilters { get; init; }

    public List<FilterDefinition> Filters { get; } = [];

    public List<FilterDefinition> ShellFilters { get; } = [];

    public List<FilterDefinition> TabLocalFilters { get; } = [];

    public List<string> DashboardFilters { get; } = [];

    public List<TabDefinition> Tabs { get; } = [];

    public List<CardDefinition> Cards { get; } = [];

    public string? ConnectorId { get; set; }

    public string? ColorPalette { get; set; }

    public LayoutDefinition Layout { get; set; } = LayoutDefinition.Default;

    public FiltersChromeDefinition FiltersChrome { get; set; } = FiltersChromeDefinition.Default;

    public LayoutBoardDefinition? LayoutBoard { get; set; }

    public LayoutBoardDefinition? ToolbarBoard { get; set; }

    public string? TabModuleLabel { get; set; }

    public ModuleIncludeState Includes { get; init; } = new();

    public DashSpecParseOptions ParseOptions { get; init; } = DashSpecParseOptions.Default;

    public string? CurrentPhaseId { get; set; }

    public string? CurrentPageId { get; set; }

    public List<ReportPageDefinition> Pages { get; } = [];

    public ModuleExtensionsDefinition ModuleExtensions { get; set; } = ModuleExtensionsDefinition.Empty;

    public Dictionary<string, string> CommandAliases { get; } = new(StringComparer.OrdinalIgnoreCase);

    public IReadOnlyList<FilterDefinition> CardBindValidationFilters =>
        MergeFilterScopes(ParentFilters, ShellFilters, TabLocalFilters, Filters);

    public IReadOnlyList<FilterDefinition> ExportedTabLocalFilters =>
        Mode is DashboardShellMode.TabModuleEmbedded ? TabLocalFilters : [];

    public static IReadOnlyList<FilterDefinition> MergeFilterScopes(
        IReadOnlyList<FilterDefinition>? parent,
        IReadOnlyList<FilterDefinition> shell,
        params IReadOnlyList<FilterDefinition>[] additional)
    {
        var merged = new Dictionary<string, FilterDefinition>(StringComparer.OrdinalIgnoreCase);
        if (parent is not null)
        {
            foreach (var filter in parent)
            {
                merged[filter.Name] = filter;
            }
        }

        foreach (var filter in shell)
        {
            merged[filter.Name] = filter;
        }

        foreach (var scope in additional)
        {
            foreach (var filter in scope)
            {
                merged[filter.Name] = filter;
            }
        }

        return merged.Values.ToList();
    }

    public void AssignToolbarBoard(LayoutBoardDefinition board, string context)
    {
        if (ToolbarBoard is not null)
        {
            throw new DashSpecParseException($"{context} declares more than one toolbar layout board.");
        }

        LayoutModuleScopeValidator.EnsureMatchesIncludeSite(board, LayoutScope.Toolbar, context);
        ToolbarBoard = board;
    }

    public void AssignTabLayoutBoard(LayoutBoardDefinition board, string context)
    {
        if (LayoutBoard is not null)
        {
            throw new DashSpecParseException($"{context} declares more than one card layout board.");
        }

        LayoutModuleScopeValidator.EnsureMatchesIncludeSite(board, LayoutScope.Tab, context);
        LayoutBoard = board;
    }
}

internal static class DashboardShellParser
{
    public static bool TryParseStatement(TokenReader reader, DashboardShellContext ctx)
    {
        if (ctx.Mode is DashboardShellMode.DashboardBody)
        {
            if (TryParseIncludeToolbar(reader, ctx))
            {
                return true;
            }
        }

        if (ctx.Mode is DashboardShellMode.TabModuleStandalone or DashboardShellMode.TabModuleEmbedded)
        {
            if (TryParseIncludeLayout(reader, ctx))
            {
                return true;
            }
        }

        if (reader.TryKeyword("connector"))
        {
            var connectorId = reader.ReadIdent();
            if (ctx.Mode is DashboardShellMode.DashboardBody or DashboardShellMode.TabModuleStandalone)
            {
                ctx.ConnectorId = connectorId;
            }

            reader.SkipNewlines();
            return true;
        }

        if (reader.TryKeyword("layout"))
        {
            var grid = LayoutParser.ParseGrid(reader);
            if (ctx.Mode is DashboardShellMode.DashboardBody or DashboardShellMode.TabModuleStandalone)
            {
                ctx.Layout = grid;
            }

            reader.SkipNewlines();
            return true;
        }

        if (reader.TryKeyword("palette"))
        {
            var palette = DashboardParser.ReadPaletteReference(reader);
            if (ctx.Mode is DashboardShellMode.DashboardBody or DashboardShellMode.TabModuleStandalone)
            {
                ctx.ColorPalette = palette;
            }

            reader.SkipNewlines();
            return true;
        }

        if (reader.TryKeyword("filters") || reader.TryKeyword("toolbar"))
        {
            ParseFiltersChrome(reader, ctx);
            reader.SkipNewlines();
            return true;
        }

        if (reader.TryKeyword("commands"))
        {
            foreach (var (alias, filterId) in CommandAliasesParser.Parse(reader))
            {
                ctx.CommandAliases[alias] = filterId;
            }

            reader.SkipNewlines();
            return true;
        }

        if (reader.TryKeyword("filter"))
        {
            var filter = FilterParser.Parse(reader);
            if (ctx.Mode is DashboardShellMode.TabModuleEmbedded)
            {
                ctx.ShellFilters.Add(filter);
            }
            else
            {
                ctx.Filters.Add(filter);
            }

            reader.SkipNewlines();
            return true;
        }

        if (reader.TryKeyword("tab"))
        {
            ParseTabStatement(reader, ctx);
            reader.SkipNewlines();
            return true;
        }

        if (reader.TryKeyword("phase"))
        {
            var phaseId = reader.ReadIdent();
            if (string.IsNullOrWhiteSpace(phaseId))
            {
                throw new DashSpecParseException("phase requires id.");
            }

            BlockSyntax.BeginBlock(reader);
            reader.SkipNewlines();
            var previousPhase = ctx.CurrentPhaseId;
            ctx.CurrentPhaseId = phaseId;
            while (!BlockSyntax.IsBlockEnd(reader, "phase", phaseId) && !reader.IsEof)
            {
                reader.SkipNewlines();
                if (BlockSyntax.IsBlockEnd(reader, "phase", phaseId))
                {
                    break;
                }

                if (!TryParseStatement(reader, ctx))
                {
                    throw reader.Unexpected();
                }
            }

            BlockSyntax.ExpectBlockEnd(reader, "phase", phaseId);
            ctx.CurrentPhaseId = previousPhase;
            reader.SkipNewlines();
            return true;
        }

        if (reader.TryKeyword("card"))
        {
            ctx.Cards.Add(CardParser.Parse(
                reader,
                ctx.CardBindValidationFilters,
                ctx.SpecDirectory,
                ctx.Includes,
                ctx.ParseOptions,
                ctx.CurrentPhaseId,
                ctx.CurrentPageId));
            reader.SkipNewlines();
            return true;
        }

        return false;
    }

    private static void ParseFiltersChrome(TokenReader reader, DashboardShellContext ctx)
    {
        ParseFiltersChrome(reader, ctx, assign: ctx.Mode is DashboardShellMode.DashboardBody or DashboardShellMode.TabModuleStandalone);
    }

    internal static void ParseFiltersChromePublic(TokenReader reader, DashboardShellContext ctx, bool assign) =>
        ParseFiltersChrome(reader, ctx, assign);

    private static void ParseFiltersChrome(TokenReader reader, DashboardShellContext ctx, bool assign)
    {
        var assignChrome = assign;

        if (reader.TryKeywordSameLine("dashboard"))
        {
            if (assignChrome)
            {
                ToolbarPlacementParser.Parse(reader, ctx, "filters dashboard");
            }
            else
            {
                ToolbarPlacementParser.Discard(reader, "filters dashboard");
            }

            return;
        }

        if (reader.TryKeywordSameLine("chrome"))
        {
            var chrome = FiltersChromeParser.Parse(reader);
            if (assignChrome)
            {
                ctx.FiltersChrome = chrome;
            }

            return;
        }

        if (assignChrome)
        {
            ToolbarPlacementParser.Parse(reader, ctx, "toolbar");
        }
        else
        {
            ToolbarPlacementParser.Discard(reader, "toolbar");
        }
    }

    private static bool TryParseIncludeToolbar(TokenReader reader, DashboardShellContext ctx)
    {
        if (!reader.TryKeyword("include"))
        {
            return false;
        }

        var (kind, reference) = DiagramModuleParser.ReadIncludeReference(reader);
        if (!string.Equals(kind, "toolbar", StringComparison.OrdinalIgnoreCase))
        {
            throw new DashSpecParseException(
                $"Dashboard shell allows include toolbar only, got include {kind}.");
        }

        if (string.IsNullOrWhiteSpace(ctx.SpecDirectory))
        {
            throw new DashSpecParseException(
                "include toolbar requires spec directory when parsing (path to the .dashspec folder).");
        }

        ctx.AssignToolbarBoard(LayoutModuleParser.Load(reference, ctx.SpecDirectory), "include toolbar");
        reader.SkipNewlines();
        return true;
    }

    private static void ParseTabStatement(TokenReader reader, DashboardShellContext ctx)
    {
        switch (ctx.Mode)
        {
            case DashboardShellMode.DashboardBody:
                ctx.Tabs.Add(TabParser.Parse(reader));
                return;

            case DashboardShellMode.TabModuleStandalone or DashboardShellMode.TabModuleEmbedded:
            {
                if (string.IsNullOrWhiteSpace(ctx.TabModuleId))
                {
                    throw new DashSpecParseException("Tab module shell requires @tab id before tab block.");
                }

                var (moduleLabel, moduleFilters, moduleLayout) =
                    TabParser.ParseModuleLocalBlock(reader, ctx.TabModuleId, allowFilters: true);
                ApplyTabModuleBlock(ctx, moduleLabel, moduleFilters, moduleLayout);
                return;
            }

            default:
                throw new DashSpecParseException("Unexpected tab statement in dashboard shell.");
        }
    }

    private static void ApplyTabModuleBlock(
        DashboardShellContext ctx,
        string? moduleLabel,
        IReadOnlyList<FilterDefinition> moduleFilters,
        LayoutBoardDefinition? moduleLayout)
    {
        ctx.TabModuleLabel ??= moduleLabel;

        if (moduleLayout is not null && ctx.LayoutBoard is not null)
        {
            throw new DashSpecParseException(
                $"Tab module '{ctx.TabModuleId}' declares layout twice: include layout and tab {{ layout }}.");
        }

        ctx.LayoutBoard ??= moduleLayout;

        if (ctx.Mode is DashboardShellMode.TabModuleStandalone)
        {
            ctx.Filters.AddRange(moduleFilters);
            return;
        }

        ctx.TabLocalFilters.AddRange(moduleFilters);
    }

    private static bool TryParseIncludeLayout(TokenReader reader, DashboardShellContext ctx)
    {
        if (!reader.TryKeyword("include"))
        {
            return false;
        }

        var (kind, reference) = DiagramModuleParser.ReadIncludeReference(reader);
        if (!string.Equals(kind, "layout", StringComparison.OrdinalIgnoreCase))
        {
            throw new DashSpecParseException(
                $"Tab module shell allows include layout only at file top, got include {kind}.");
        }

        if (string.IsNullOrWhiteSpace(ctx.SpecDirectory))
        {
            throw new DashSpecParseException(
                "include layout requires spec directory when parsing (path to the .dashspec folder).");
        }

        if (ctx.LayoutBoard is not null)
        {
            throw new DashSpecParseException("Tab module declares more than one layout board.");
        }

        ctx.AssignTabLayoutBoard(LayoutModuleParser.Load(reference, ctx.SpecDirectory), "include layout");
        reader.SkipNewlines();
        return true;
    }
}

File: src/DashSpec.Modeling.Parse/Document/DashboardShellParser.fs
namespace DashSpec.Modeling.Parse.Document

open System
open System.Collections.Generic
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Card
open DashSpec.Modeling.Parse.Diagram
open DashSpec.Modeling.Parse.Filter
open DashSpec.Modeling.Parse.Layout
open DashSpec.Modeling.Parse.Lexing
open DashSpec.Modeling.Parse.Toolbar

module DashboardShellParser =

    let private toolbarContext (ctx: DashboardShellContext) (blockName: string) =
        { AssignToolbarBoard = fun board -> ctx.AssignToolbarBoard(board, blockName)
          AddDashboardFilters = fun names -> ctx.DashboardFilters.AddRange names }

    let private parseFiltersChrome (reader: TokenReader) (ctx: DashboardShellContext) (assign: bool) =
        if reader.TryKeywordSameLine "dashboard" then
            if assign then
                ToolbarPlacementParser.parse reader (toolbarContext ctx "filters dashboard") "filters dashboard"
            else
                ToolbarPlacementParser.discard reader "filters dashboard"
        elif reader.TryKeywordSameLine "chrome" then
            let chrome = FiltersChromeParser.parse reader
            if assign then ctx.FiltersChrome <- chrome
        elif assign then
            ToolbarPlacementParser.parse reader (toolbarContext ctx "toolbar") "toolbar"
        else
            ToolbarPlacementParser.discard reader "toolbar"

    let parseFiltersChromePublic (reader: TokenReader) (ctx: DashboardShellContext) (assign: bool) =
        parseFiltersChrome reader ctx assign

    let private tryParseIncludeToolbar (reader: TokenReader) (ctx: DashboardShellContext) =
        if not (reader.TryKeyword "include") then false
        else
            let kind, reference = DiagramModuleParser.readIncludeReference reader
            if not (String.Equals(kind, "toolbar", StringComparison.OrdinalIgnoreCase)) then
                raise (DashSpecParseException($"Dashboard shell allows include toolbar only, got include {kind}."))
            match ctx.SpecDirectory with
            | None | Some "" ->
                raise (DashSpecParseException("include toolbar requires spec directory when parsing (path to the .dashspec folder)."))
            | Some specDirectory ->
                ctx.AssignToolbarBoard(LayoutModuleParser.load reference specDirectory, "include toolbar")
                reader.SkipNewlines()
                true

    let private tryParseIncludeLayout (reader: TokenReader) (ctx: DashboardShellContext) =
        if not (reader.TryKeyword "include") then false
        else
            let kind, reference = DiagramModuleParser.readIncludeReference reader
            if not (String.Equals(kind, "layout", StringComparison.OrdinalIgnoreCase)) then
                raise (DashSpecParseException($"Tab module shell allows include layout only at file top, got include {kind}."))
            match ctx.SpecDirectory with
            | None | Some "" ->
                raise (DashSpecParseException("include layout requires spec directory when parsing (path to the .dashspec folder)."))
            | Some specDirectory ->
                if ctx.LayoutBoard.IsSome then
                    raise (DashSpecParseException("Tab module declares more than one layout board."))
                ctx.AssignTabLayoutBoard(LayoutModuleParser.load reference specDirectory, "include layout")
                reader.SkipNewlines()
                true

    let private applyTabModuleBlock
        (ctx: DashboardShellContext)
        (moduleLabel: string option)
        (moduleFilters: IReadOnlyList<FilterDefinition>)
        (moduleLayout: LayoutBoardDefinition option)
        =
        ctx.TabModuleLabel <- ctx.TabModuleLabel |> Option.orElse moduleLabel

        if moduleLayout.IsSome && ctx.LayoutBoard.IsSome then
            raise (DashSpecParseException($"Tab module '{ctx.TabModuleId}' declares layout twice: include layout and tab {{ layout }}."))

        ctx.LayoutBoard <- ctx.LayoutBoard |> Option.orElse moduleLayout

        match ctx.Mode with
        | DashboardShellMode.TabModuleStandalone -> ctx.Filters.AddRange moduleFilters
        | DashboardShellMode.TabModuleEmbedded -> ctx.TabLocalFilters.AddRange moduleFilters
        | _ -> ()

    let private parseTabStatement (reader: TokenReader) (ctx: DashboardShellContext) =
        match ctx.Mode with
        | DashboardShellMode.DashboardBody -> ctx.Tabs.Add(TabParser.parse reader)
        | DashboardShellMode.TabModuleStandalone
        | DashboardShellMode.TabModuleEmbedded ->
            match ctx.TabModuleId with
            | None | Some "" -> raise (DashSpecParseException("Tab module shell requires @tab id before tab block."))
            | Some tabModuleId ->
                let moduleLabel, moduleFilters, moduleLayout =
                    TabParser.parseModuleLocalBlock reader tabModuleId true
                applyTabModuleBlock ctx moduleLabel moduleFilters moduleLayout

    let rec tryParseStatement (reader: TokenReader) (ctx: DashboardShellContext) =
        if ctx.Mode = DashboardShellMode.DashboardBody && tryParseIncludeToolbar reader ctx then
            true
        elif (ctx.Mode = DashboardShellMode.TabModuleStandalone || ctx.Mode = DashboardShellMode.TabModuleEmbedded)
             && tryParseIncludeLayout reader ctx then
            true
        elif reader.TryKeyword "connector" then
            let connectorId = reader.ReadIdent()
            if ctx.Mode = DashboardShellMode.DashboardBody || ctx.Mode = DashboardShellMode.TabModuleStandalone then
                ctx.ConnectorId <- Some connectorId
            reader.SkipNewlines()
            true
        elif reader.TryKeyword "layout" then
            let grid = LayoutParser.parseGrid reader
            if ctx.Mode = DashboardShellMode.DashboardBody || ctx.Mode = DashboardShellMode.TabModuleStandalone then
                ctx.Layout <- grid
            reader.SkipNewlines()
            true
        elif reader.TryKeyword "palette" then
            let palette =
                if reader.RawKind = TokenKind.Eq then
                    reader.Advance()
                reader.ReadScalarValue()
            if ctx.Mode = DashboardShellMode.DashboardBody || ctx.Mode = DashboardShellMode.TabModuleStandalone then
                ctx.ColorPalette <- Some palette
            reader.SkipNewlines()
            true
        elif reader.TryKeyword "filters" || reader.TryKeyword "toolbar" then
            let assign =
                ctx.Mode = DashboardShellMode.DashboardBody || ctx.Mode = DashboardShellMode.TabModuleStandalone
            parseFiltersChrome reader ctx assign
            reader.SkipNewlines()
            true
        elif reader.TryKeyword "commands" then
            for KeyValue(alias, filterId) in CommandAliasesParser.parse reader do
                ctx.CommandAliases.[alias] <- filterId
            reader.SkipNewlines()
            true
        elif reader.TryKeyword "filter" then
            let filter = FilterParser.parse reader
            if ctx.Mode = DashboardShellMode.TabModuleEmbedded then
                ctx.ShellFilters.Add filter
            else
                ctx.Filters.Add filter
            reader.SkipNewlines()
            true
        elif reader.TryKeyword "tab" then
            parseTabStatement reader ctx
            reader.SkipNewlines()
            true
        elif reader.TryKeyword "phase" then
            let phaseId = reader.ReadIdent()
            if String.IsNullOrWhiteSpace phaseId then
                raise (DashSpecParseException("phase requires id."))
            BlockSyntax.beginBlock reader
            reader.SkipNewlines()
            let previousPhase = ctx.CurrentPhaseId
            ctx.CurrentPhaseId <- Some phaseId
            while not (BlockSyntax.isBlockEnd reader "phase" (Some phaseId)) && not reader.IsEof do
                reader.SkipNewlines()
                if BlockSyntax.isBlockEnd reader "phase" (Some phaseId) then ()
                elif not (tryParseStatement reader ctx) then
                    raise (reader.Unexpected())
            BlockSyntax.expectBlockEnd reader "phase" (Some phaseId)
            ctx.CurrentPhaseId <- previousPhase
            reader.SkipNewlines()
            true
        elif reader.TryKeyword "card" then
            ctx.Cards.Add(
                CardParser.parse
                    reader
                    ctx.CardBindValidationFilters
                    ctx.SpecDirectory
                    (Some ctx.Includes)
                    (Some ctx.ParseOptions)
                    ctx.CurrentPhaseId
                    ctx.CurrentPageId)
            reader.SkipNewlines()
            true
        else
            false
